pep-0567.rst

```diff
@@ -461,37 +487,64 @@ points to a ``_ContextData`` object::
             return self._old_value
 
 
-Implementation Notes
-====================
+Summary of the New APIs
+=======================
+
+Python API
+----------
 
-* The internal immutable dictionary for ``Context`` is implemented
-  using Hash Array Mapped Tries (HAMT).  They allow for O(log N)
-  ``set`` operation, and for O(1) ``copy_context()`` function, where
-  *N* is the number of items in the dictionary.  For a detailed
-  analysis of HAMT performance please refer to :pep:`550` [1]_.
+1. A new ``contextvars`` module with ``ContextVar``, ``Context``,
+   and ``Token`` classes, and a ``copy_context()`` function.
 
-* ``ContextVar.get()`` has an internal cache for the most recent
-  value, which allows to bypass a hash lookup.  This is similar
-  to the optimization the ``decimal`` module implements to
-  retrieve its context from ``PyThreadState_GetDict()``.
-  See :pep:`550` which explains the implementation of the cache
-  in a great detail.
+2. ``asyncio.Loop.call_at()``, ``asyncio.Loop.call_later()``,
+   ``asyncio.Loop.call_soon()``, and
+   ``asyncio.Future.add_done_callback()`` run callback functions in
+   the context they were called in.  A new *context* keyword-only
+   parameter can be used to specify a custom context.
 
+3. ``asyncio.Task`` is modified internally to maintain its own
+   context.
 
-Summary of the New APIs
-=======================
 
-* A new ``contextvars`` module with ``ContextVar``, ``Context``,
-  and ``Token`` classes, and a ``copy_context()`` function.
+C API
+-----
+
+1. ``PyContextVar * PyContextVar_New(char *name, PyObject *default)``:
+   create a ``ContextVar`` object.  The *default* argument can be
+   ``NULL``, which means that the variable has no default value.
+
+2. ``int PyContextVar_Get(PyContextVar *, PyObject *default_value, PyObject **value)``:
+   return ``-1`` if an error occurs during the lookup, ``0`` otherwise.
+   If a value for the context variable is found, it will be set to the
+   ``value`` pointer.  Otherwise, ``value`` will be set to
+   ``default_value`` when it is not ``NULL``.  If ``default_value`` is
+   ``NULL``, ``value`` will be set to the default value of the
+   variable, which can be ``NULL`` too.  ``value`` is always a new
+   reference.
+
+3. ``PyContextToken * PyContextVar_Set(PyContextVar *, PyObject *)``:
+   set the value of the variable in the current context.
+
+4. ``PyContextVar_Reset(PyContextVar *, PyContextToken *)``:
+   reset the value of the context variable.
+
+5. ``PyContext * PyContext_New()``: create a new empty context.
+
+6. ``PyContext * PyContext_Copy()``: get a copy of the current context.
+
+7. ``int PyContext_Enter(PyContext *)`` and
+   ``int PyContext_Exit(PyContext *)`` allow to set and restore
+   the context for the current OS thread.  It is required to always
+   restore the previous context::
```

Comment:
This requirement has no teeth -- you cannot enforce it, can you?
Also, someone on the mailing list (Paul?) asked why we need separate Enter and Exit functions -- is this answered anywhere?

Reply:
> This requirement has no teeth -- you cannot enforce it, can you?

No, it's not enforceable, but I think we should document it that way.  Otherwise `PyContext_Exit` will get out of order, which will in turn result in errors.

> Also, someone on the mailing list (Paul?) asked why we need separate Enter and Exit functions -- is this answered anywhere?

This is just how we usually do this in C.  We could design an API that would accept a pointer to a function and run it within the context, but such API would be painful to use.  Hence enter/exit methods.